File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/dao/NoticeDao.java
/*
 * Copyright (c) Kacper Ziubryniewicz 2020-1-6
 */

package pl.szczodrzynski.edziennik.data.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.RawQuery;
import androidx.sqlite.db.SimpleSQLiteQuery;
import androidx.sqlite.db.SupportSQLiteQuery;

import java.util.List;

import pl.szczodrzynski.edziennik.data.db.entity.Metadata;
import pl.szczodrzynski.edziennik.data.db.entity.Notice;
import pl.szczodrzynski.edziennik.data.db.full.NoticeFull;

import static pl.szczodrzynski.edziennik.data.db.entity.Metadata.TYPE_NOTICE;

@Dao
public abstract class NoticeDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract long add(Notice notice);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void addAll(List<Notice> noticeList);

    @Query("DELETE FROM notices WHERE profileId = :profileId")
    public abstract void clear(int profileId);

    @Query("DELETE FROM notices WHERE profileId = :profileId AND noticeSemester = :semester")
    public abstract void clearForSemester(int profileId, int semester);

    @RawQuery(observedEntities = {Notice.class})
    abstract LiveData<List<NoticeFull>> getAll(SupportSQLiteQuery query);
    public LiveData<List<NoticeFull>> getAll(int profileId, String filter) {
        return getAll(new SimpleSQLiteQuery("SELECT \n" +
                "*, \n" +
                "teachers.teacherName || ' ' || teachers.teacherSurname AS teacherFullName\n" +
                "FROM notices \n" +
                "LEFT JOIN teachers USING(profileId, teacherId)\n" +
                "LEFT JOIN metadata ON noticeId = thingId AND thingType = "+TYPE_NOTICE+" AND metadata.profileId = "+profileId+"\n" +
                "WHERE notices.profileId = "+profileId+" AND "+filter+"\n" +
                "ORDER BY addedDate DESC"));
    }
    public LiveData<List<NoticeFull>> getAll(int profileId) {
        return getAll(profileId, "1");
    }
    public LiveData<List<NoticeFull>> getAllWhere(int profileId, String filter) {
        return getAll(profileId, filter);
    }

    @RawQuery(observedEntities = {Notice.class, Metadata.class})
    abstract List<NoticeFull> getAllNow(SupportSQLiteQuery query);
    public List<NoticeFull> getAllNow(int profileId, String filter) {
        return getAllNow(new SimpleSQLiteQuery("SELECT \n" +
                "*, \n" +
                "teachers.teacherName || ' ' || teachers.teacherSurname AS teacherFullName\n" +
                "FROM notices \n" +
                "LEFT JOIN teachers USING(profileId, teacherId)\n" +
                "LEFT JOIN metadata ON noticeId = thingId AND thingType = "+TYPE_NOTICE+" AND metadata.profileId = "+profileId+"\n" +
                "WHERE notices.profileId = "+profileId+" AND "+filter+"\n" +
                "ORDER BY addedDate DESC"));
    }
    public List<NoticeFull> getNotNotifiedNow(int profileId) {
        return getAllNow(profileId, "notified = 0");
    }

    @Query("SELECT " +
            "*, " +
            "teachers.teacherName || ' ' || teachers.teacherSurname AS teacherFullName " +
            "FROM notices " +
            "LEFT JOIN teachers USING(profileId, teacherId) " +
            "LEFT JOIN metadata ON noticeId = thingId AND thingType = "+TYPE_NOTICE+" AND metadata.profileId = notices.profileId " +
            "WHERE notified = 0 " +
            "ORDER BY addedDate DESC")
    public abstract List<NoticeFull> getNotNotifiedNow();
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/dao/NoticeDao.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2020-4-25.
 */
package pl.szczodrzynski.edziennik.data.db.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Query
import androidx.room.RawQuery
import androidx.sqlite.db.SupportSQLiteQuery
import eu.szkolny.selectivedao.annotation.SelectiveDao
import eu.szkolny.selectivedao.annotation.UpdateSelective
import pl.szczodrzynski.edziennik.App
import pl.szczodrzynski.edziennik.data.db.AppDb
import pl.szczodrzynski.edziennik.data.db.entity.Metadata
import pl.szczodrzynski.edziennik.data.db.entity.Notice
import pl.szczodrzynski.edziennik.data.db.full.NoticeFull

@Dao
@SelectiveDao(db = AppDb::class)
abstract class NoticeDao : BaseDao<Notice, NoticeFull> {
    companion object {
        private const val QUERY = """
            SELECT 
            *, 
            teachers.teacherName ||" "|| teachers.teacherSurname AS teacherName
            FROM notices
            LEFT JOIN teachers USING(profileId, teacherId)
            LEFT JOIN metadata ON noticeId = thingId AND thingType = 2 AND metadata.profileId = notices.profileId
        """

        private const val ORDER_BY = """ORDER BY addedDate DESC"""
    }

    private val selective by lazy { NoticeDaoSelective(App.db) }

    @RawQuery(observedEntities = [Notice::class])
    abstract override fun getRaw(query: SupportSQLiteQuery): LiveData<List<NoticeFull>>
    @RawQuery(observedEntities = [Notice::class])
    abstract override fun getOne(query: SupportSQLiteQuery): LiveData<NoticeFull?>

    // SELECTIVE UPDATE
    @UpdateSelective(primaryKeys = ["profileId", "noticeId"], skippedColumns = ["addedDate"])
    override fun update(item: Notice) = selective.update(item)
    override fun updateAll(items: List<Notice>) = selective.updateAll(items)

    // CLEAR
    @Query("DELETE FROM notices WHERE profileId = :profileId")
    abstract override fun clear(profileId: Int)
    // REMOVE NOT KEPT
    @Query("DELETE FROM notices WHERE keep = 0")
    abstract override fun removeNotKept()

    // GET ALL - LIVE DATA
    fun getAll(profileId: Int) =
            getRaw("$QUERY WHERE notices.profileId = $profileId $ORDER_BY")

    // GET ALL - NOW
    fun getAllNow(profileId: Int) =
            getRawNow("$QUERY WHERE notices.profileId = $profileId $ORDER_BY")
    fun getNotNotifiedNow() =
            getRawNow("$QUERY WHERE notified = 0 $ORDER_BY")
    fun getNotNotifiedNow(profileId: Int) =
            getRawNow("$QUERY WHERE notices.profileId = $profileId AND notified = 0 $ORDER_BY")

    // GET ONE - NOW
    fun getByIdNow(profileId: Int, id: Long) =
            getOneNow("$QUERY WHERE notices.profileId = $profileId AND noticeId = $id")

    @Query("UPDATE notices SET keep = 0 WHERE profileId = :profileId AND noticeSemester = :semester")
    abstract fun dontKeepSemester(profileId: Int, semester: Int)
}
